### ISSUE TITLE
"$refs must reference a valid location in the document" issue with schema that has both allOf and properties

### ISSUE THREAD

**Issue by jmarikle** [zircote/swagger-php#952]
# The Issue
We are currently upgrading from 3.0.2 to 3.2.1 and running into the above error.  The reason this is occurring is because we reference properties of schemas in our documentation (e.g. `#/components/schemas/BaseModel/properties/id`), but after upgrading it's breaking with a message similar to the following:

> $ref "#/components/schemas/\<SchemaName\>/properties/\<PropertyName\>" not found for @OA\Property(property="\<PropertyName\>") in \</path/to/included/file\>.php on line 11

# The Example
Here's a quick, simplified example that works on https://editor.swagger.io/ but does not work in our application.
```
openapi: 3.0.0
info:
  title: 'Demo for possible swagger-php issue'
  version: 1.0.0
paths:
  '/products/{product_id}':
    post:
      tags:
        - Products
      operationId: App\\Http\\Controllers\\ProductsController::index
      parameters:
        -
          name: product_id
          in: path
          description: 'ID of product to update'
          required: true
          schema:
            $ref: '#/components/schemas/BaseModel/properties/id'
      requestBody:
        content:
          application/json:
            schema:
               $ref: '#/components/schemas/UpdateProductRequest'
      responses:
        200:
          description: 'successful operation'
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Product'
components:
  schemas:
    BaseModel:
      title: 'Base model'
      properties:
        id:
          description: 'The unique identifier of a model.'
          type: integer
          format: int64
          example: 1
        created_at:
          description: 'Timestamp of when the model was created.'
          type: string
          format: datetime
          readOnly: true
          example: "2020-07-07 20:20:20"
        updated_at:
          description: 'Timestamp of when the model was updated.'
          type: string
          format: datetime
          readOnly: true
          example: "2020-07-07 20:20:20"
    Product:
      title: 'Product model'
      allOf:
        -
          $ref: '#/components/schemas/BaseModel'
      properties:
        price:
          description: 'The product price in micro amounts.'
          type: integer
          format: int64
          example: 9990000
    UpdateProductRequest:
      title: 'Create Product Request'
      properties:
        price:
          $ref: '#/components/schemas/Product/properties/price'
```

# The Reason
I believe the above is now breaking in our project as a result of the following change: https://github.com/zircote/swagger-php/commit/472b56d94bf7dedc7673431d0a194d7d9a046cbc#diff-31ba58b4eef5715dba10859c144fcd63bb0e239a8d05f91d151d34230a4c3236R91-R109.  It takes the properties we were referencing and places them inside the `allOf` declaration if it exists for that schema.  This breaks the reference to that property elsewhere.  In the example you can see that we use `$ref: '#/components/schemas/Product/properties/price'` on line 68, which works because it's a direct property of `components/schemas/Product`.  Note that we also use `$ref: '#/components/schemas/BaseModel/properties/id'` on line 18.  We *don't* use `#/components/schemas/Product/properties/id'` because we can't (as far as I know) reference properties added through `allOf`.  Thus the issue: all properties are now added to `allOf` after the above change and break existing references.

Because the above works on https://editor.swagger.io/ but not in our codebase when we use v3.2.1, I assume this is a bug.  If it's not a bug and is in fact according to spec, do you have any advice on how to amend our definitions?

Thank you.

**Comment by DerManoMann** [comment]
Looks like a bug :/

I hope you can live with 3.0.2 for a while longer as this will need some time. The merging/inheritance stuff is on my list already as there are more edge cases that should be fixed..

**Comment by jmarikle** [comment]
Thank you for looking into this.  We currently have a temporary workaround in place.

**Comment by DerManoMann** [comment]
Hi there,

I think I've got a fix for this problem in #974 - would you be able to try that branch and let me know?

**Comment by DerManoMann** [comment]
Fixed via #974 974
